### ISSUE TITLE
training: clean BC warm-start PPO rerun from preserved artifacts

### ISSUE THREAD

**Issue by ll7** [ll7/robot_sf_ll7#1977]
Parent: #1108
Follow-up from: #1961

## Goal / Problem

Issue #1961 classified the historical #1108 / #749 BC warm-start artifact trail as `rerun_required`: the 141-episode dataset and BC checkpoint from job `12472` have a durable W&B preservation pointer, but the final PPO fine-tuned checkpoint and policy-analysis comparison were not recovered.

Run a clean, current BC warm-start PPO continuation or rerun that does not depend on stale worktree-local `output/` paths.

## Known durable inputs

- Preserved artifact: `ll7/robot_sf/issue_1108_bc_warm_start_job12472:v0`
- Aliases: `issue-1108`, `job-12472`, `preserved-20260525`
- Dataset: `issue_749_b60iopxt_v10_eval_trajectories.npz`, size `3.25 GB`, sha256 `fc62311e2dfb0cbc6742e745e7deb0ae876bda5c588a35101812d3297c902e81`
- BC checkpoint: `issue_749_bc_preinit_v10_policy.zip`, size `41.4 MB`, sha256 `c7ee44796f73c3e58dbf9ba7b006e56452b15d4a4d2dde1bc148f9ea2d826ac1`
- Preservation run: https://wandb.ai/ll7/robot_sf/runs/19udjzki

## Required output

- Hydrate or explicitly reference the preserved dataset and BC checkpoint through durable pointers.
- Run or resume the PPO fine-tune using current config contracts.
- Publish a durable final PPO checkpoint pointer, or record a fail-closed unsuccessful run.
- Run policy-analysis comparison against `27dbe5xu` and `b60iopxt`.
- Record whether warm start improved sample efficiency, final score, both, or neither.

## Out of scope

- Treating local `output/` files as durable evidence.
- Reward or hyperparameter redesign beyond the BC warm-start rerun contract.
- Paper-facing claims without the comparison evidence above.

## Validation

- Preflight artifact hydration and checksum for the preserved dataset and BC checkpoint.
- Canonical PPO training or fine-tune command with config provenance.
- Policy-analysis command and compact durable report.
- Artifact classification before closing.

**Comment by ll7** [comment]
SLURM training lane update for #1977:

- Selected this as the next `goal-slurm-experiment` candidate after the skill-owned occupancy gate was idle (`squeue --me` had no active `gse-` job). I treated #1470 as preservation/closeout work rather than a fresh training candidate, and #1475/#1472/#2006 still need readiness repair before nominal training escalation.
- Prepared branch `issue-1977-bc-warm-start-rerun` from a sibling worktree at commit `a8aeae0f` (`fix: avoid srun handshake in issue 1977 job`). No PR opened yet; per current workflow preference, this should wait until the job finishes and results are preserved.
- Added the config-first rerun surface:
  - `configs/training/ppo_imitation/ppo_finetune_issue_1977_bc_warm_start_rerun.yaml`
  - `SLURM/Auxme/issue_1977_bc_warm_start_ppo.sl`
  - `scripts/dev/sbatch_issue1977_bc_warm_start_ppo.sh`
- Preflight evidence before submission:
  - `bash -n scripts/dev/sbatch_issue1977_bc_warm_start_ppo.sh SLURM/Auxme/issue_1977_bc_warm_start_ppo.sl`
  - `git diff --check`
  - `uv run python scripts/training/train_ppo_with_pretrained_policy.py --config configs/training/ppo_imitation/ppo_finetune_issue_1977_bc_warm_start_rerun.yaml --dry-run`
  - W&B artifact `ll7/robot_sf/issue_1108_bc_warm_start_job12472:v0` exists with digest `23671891594563fb04a6ddc800c77893` and the expected checkpoint/dataset/manifest files.
- Job `12686` failed in 17s from the known Auxme `srun` allocation handshake path (`Unable to confirm allocation... Zero Bytes were transmitted or received`). I classified that as wrapper infrastructure failure, not model/training failure, and fixed the launcher to run directly inside the batch allocation by default.
- Active replacement job: `12687` / `gse-1977-bcppo`, partition `a30`, node `auxme-imech172`, started `2026-06-01T14:04:10`, walltime `1-12:00:00`, log `output/slurm/12687-issue1977-bc-warm-start-ppo.out`.
- Early runtime evidence from job `12687`:
  - W&B artifact download completed.
  - Hydrated checkpoint checksum: `c7ee44796f73c3e58dbf9ba7b006e56452b15d4a4d2dde1bc148f9ea2d826ac1`.
  - Hydrated dataset checksum: `fc62311e2dfb0cbc6742e745e7deb0ae876bda5c588a35101812d3297c902e81`.
  - PPO fine-tune has started and reached the first rollout table at `total_timesteps=2048`, `fps=59`, `success_rate=0`.

Next monitor cadence: because this is a 10M PPO fine-tune plus post-training policy analysis and #1474's 5M run took about 7h, I plan to sleep/check in long intervals rather than polling tightly: first check around 3h, then roughly 7-8h if still healthy, then near walltime only if needed. Final preservation/PR work should happen after the job exits and artifacts are inspected/promoted.

**Comment by ll7** [comment]
Runtime diagnosis update for active job `12687`:

The low throughput appears to be explained by the current PPO fine-tune path, not by a transient cluster issue.

Evidence:
- The active log explicitly says SB3 wrapped the fine-tune env in `DummyVecEnv`, so the run is using a single environment.
- `scripts/training/train_ppo_with_pretrained_policy.py` loads the BC policy with `PPO.load(..., env=env)` and then calls one uninterrupted `model.learn(total_timesteps=10000000, ...)`; it does not use the high-throughput `scripts/training/train_ppo.py` vectorized path.
- `PPOFineTuningConfig` currently has no `num_envs`, `worker_mode`, `device`, `ppo_hyperparams`, or periodic checkpoint fields, so the referenced source training config's `num_envs: auto` is not actually applied to this fine-tune run.
- The log has SB3's warning that this MLP PPO policy is running on GPU, where GPU utilization is expected to be poor.
- The config also pins `predictive_foresight_device: cpu`, while nearby high-throughput A30 promotion configs use CUDA predictive foresight and `num_envs: 22`/subprocess workers.
- At ~57 fps, 10M steps would need about 48.7h, but the a30 partition max is 36h. The job is likely to be killed around 7.3M steps.
- More importantly, this fine-tune script saves the final checkpoint only after `model.learn(...)` returns. If SLURM kills the job at walltime, this can repeat the old #1108 failure mode: useful training time but no durable final PPO checkpoint.

Recommendation before claiming this as the clean #1977 rerun:
1. Do not treat job `12687` as likely to produce the intended 10M artifact.
2. Prefer cancelling/resubmitting early after a small launcher/code correction, because the job is still near the beginning and the current path is structurally unlikely to finish cleanly.
3. Minimal safe repair: add periodic checkpointing to `train_ppo_with_pretrained_policy.py` and set a walltime-safe target if keeping the single-env path.
4. Better repair: add explicit fine-tune support for `num_envs`/`worker_mode` and a short SLURM smoke proving the BC checkpoint can be loaded with a vectorized env before submitting the 10M run again.

I have not cancelled the active job yet; this comment preserves the diagnosis and tradeoff before changing the live run.

**Comment by ll7** [comment]
Repair update for the #1977 SLURM training lane:

- Cancelled job `12687` after diagnosing the low-throughput path. It was cancelled at `2026-06-01T15:39:10+02:00` after `01:35:00`; it had been using single-env `DummyVecEnv` at about `57 fps` and was unlikely to finish 10M before walltime or save a final checkpoint.
- Implemented and pushed repair commit `f3234dc0` on branch `issue-1977-bc-warm-start-rerun`:
  - `PPOFineTuningConfig` now supports `num_envs`, `worker_mode`, `device`, PPO hyperparameter overrides, and periodic checkpoint fields.
  - `train_ppo_with_pretrained_policy.py` now builds `DummyVecEnv`/`SubprocVecEnv`, loads the BC checkpoint into that vectorized env, reapplies fine-tune hyperparameters after `PPO.load`, and installs periodic checkpointing.
  - #1977/#749 fine-tune configs now use `num_envs: 22`, `worker_mode: subproc`, `device: cpu`, `predictive_foresight_device: cuda`, and `checkpoint_freq: 500000`.
  - #1977 wrapper now accepts `--config` and `--wandb-artifact` for repeatable smoke/prod submission.
- Validation before full resubmit:
  - `uv run python -m py_compile robot_sf/training/imitation_config.py scripts/training/train_ppo_with_pretrained_policy.py`
  - `uv run ruff check robot_sf/training/imitation_config.py scripts/training/train_ppo_with_pretrained_policy.py tests/training/test_train_ppo_with_pretrained_policy.py tests/integration/test_ppo_pretraining_pipeline.py`
  - `uv run pytest tests/training/test_train_ppo_with_pretrained_policy.py tests/integration/test_ppo_pretraining_pipeline.py -q` -> `16 passed`
  - `uv run python scripts/training/train_ppo_with_pretrained_policy.py --config configs/training/ppo_imitation/ppo_finetune_issue_1977_bc_warm_start_rerun.yaml --dry-run`
  - `bash -n scripts/dev/sbatch_issue1977_bc_warm_start_ppo.sh`
- SLURM smoke `12688` (`gse-1977-smoke`) completed successfully in `00:08:18` on `a30`:
  - hydrated the same preserved W&B artifact/checksums,
  - trained `45056` timesteps at `277 fps`,
  - wrote periodic checkpoints at `22000` and `44000` steps,
  - wrote final smoke checkpoint and manifest.
- Full repaired job submitted: `12689` / `gse-1977-bcppo`, commit `f3234dc0d835e25f42181839e00ce7e9cba24e69`, `a30`, node `auxme-imech172`, start `2026-06-01T15:54:17+02:00`, walltime `1-12:00:00`, log `output/slurm/12689-issue1977-bc-warm-start-ppo.out`.
- Immediate health check for `12689`: artifact hydration succeeded and the first rollout reached `45056` timesteps at `279 fps`, so the repair is active. At this speed, 10M training is roughly 10h before post-training analysis, comfortably inside the 36h walltime.

Next monitor point: first periodic checkpoint should be around 500k timesteps, roughly 30 minutes after the first rollout at the observed speed. Final PR/artifact-preservation should still wait until job `12689` exits and outputs are inspected/promoted.

**Comment by ll7** [comment]
Stopped job `12689` per maintainer direction after the success signal stayed extremely low.

```yaml
slurm_state: cancelled_intermediate
job_id: 12689
job_name: gse-1977-bcppo
partition: a30
elapsed: 14:47:53
start: 2026-06-01T15:54:17+02:00
end: 2026-06-02T06:42:10+02:00
last_logged_total_timesteps: 5361664
last_logged_iteration: 119
last_logged_success_rate: 0
last_logged_fps: 100
last_logged_updates: 1180
```

Intermediate result boundary:
- The run is diagnostic/intermediate evidence only. Do not count it as completed #1977 evidence or as a successful BC warm-start PPO rerun.
- The local cleanup trap synced outputs to `output/slurm/issue1977-bc-warm-start-ppo-job-12689/`.
- Periodic checkpoints are present at roughly 500k intervals through `4,999,940` steps under `output/slurm/issue1977-bc-warm-start-ppo-job-12689/benchmarks/expert_policies/checkpoints/issue_1977_ppo_finetune_v10_warm_start_rerun/`.
- No final fine-tuned checkpoint/manifest or post-training policy-analysis result exists, because the job was intentionally cancelled before completion.
- These paths are local `output/` artifacts and remain non-durable until promoted or recorded through an explicit artifact pointer.

Next action: preserve/promote the intermediate checkpoints only if they are useful as negative/diagnostic evidence, then select a different SLURM candidate rather than spending more allocation on this run shape.

**Comment by ll7** [comment]
Preserved the stopped `12689` intermediate result in PR #2021.

Evidence update:
- Commit now includes `docs/context/evidence/issue_1977_bc_warm_start_cancelled_2026-06-02/artifact_manifest.md`.
- Context note `docs/context/issue_1961_bc_warm_start_recoverability.md` now records the follow-up rerun result.
- Classification: `cancelled_intermediate_negative_diagnostic`.
- Last logged state: `5,361,664` timesteps, iteration `119`, fps `100`, success rate `0`.
- Final PPO checkpoint and policy-analysis comparison are still missing because the job was intentionally cancelled.
- The periodic checkpoints through about 5M steps remain local-only diagnostic outputs, not durable promoted artifacts.

Recommendation: stop this run shape. Do not submit another 10M continuation from the same preserved inputs unless a future issue defines a learned-progress redesign or a narrow intermediate-checkpoint analysis question.

**Comment by ll7** [comment]
Closing as interpreted after PR #2021 preserved the stopped rerun evidence.\n\nResult classification: cancelled_intermediate_negative_diagnostic, not a completed BC warm-start PPO rerun and not benchmark/paper evidence. The repaired vectorized fine-tune path worked technically, but job 12689 was stopped at 5,361,664 timesteps with success_rate=0; no final PPO checkpoint or policy-analysis comparison exists.\n\nDecision: stop this run shape. Do not submit another 10M continuation from the same preserved inputs unless a future issue defines a learned-progress redesign or a narrow intermediate-checkpoint analysis question with a new stop rule. The preserved context/evidence surface is  plus .

**Comment by ll7** [comment]
Clarification for the closeout comment: the intended preserved context/evidence paths are literal repository paths:

- `docs/context/issue_1961_bc_warm_start_recoverability.md`
- `docs/context/evidence/issue_1977_bc_warm_start_cancelled_2026-06-02/artifact_manifest.md`

The closeout classification remains `cancelled_intermediate_negative_diagnostic`: technically repaired vectorized fine-tune path, stopped at 5,361,664 timesteps with success_rate=0, no final PPO checkpoint, and no policy-analysis comparison. Stop this run shape unless a future issue defines a learned-progress redesign or a narrow intermediate-checkpoint analysis question.
